File: 01_ML/01_supervised/01_regression/02_nonlinear_reg/nodejs/02a_quadratic.js
// ML: Non-Linear Regression - Polynomial / Quadratic
// (Node.js port of ../jupyter/02a_quadratic.ipynb)
//
// House age vs price: ∪-parabola — new builds priced high, mid-age homes dip,
// historic homes recover.
//
//   price = a * age^2 + b * age + c

import MLR from 'ml-regression-multivariate-linear';
import { writePlotlyHtml, linspace } from './plot.js';

const data = {
  age:   [1, 5, 10, 15, 20, 25, 30, 40, 55, 70, 85, 100],
  price: [780, 690, 600, 540, 510, 500, 510, 560, 660, 770, 880, 990],
};
const test = { age: [3, 22, 45, 90] };

console.table(data.age.map((a, i) => ({ age: a, price: data.price[i] })));

// Quadratic features: [age, age^2]
const X = data.age.map((x) => [x, x * x]);
const Y = data.price.map((p) => [p]);

const reg = new MLR(X, Y);

// weights shape: 3 x 1 → [coef_age, coef_age2, intercept]
const b = reg.weights[0][0];
const a = reg.weights[1][0];
const c = reg.weights[2][0];
console.log(`fit: price = ${a.toFixed(4)} * age^2 + ${b.toFixed(4)} * age + ${c.toFixed(4)}`);

const Xtest = test.age.map((x) => [x, x * x]);
const predictions = reg.predict(Xtest).map((row) => row[0]);
console.table(test.age.map((x, i) => ({ age: x, price: predictions[i] })));

const ageGrid = linspace(Math.min(...data.age), Math.max(...data.age), 200);
const curve = ageGrid.map((x) => reg.predict([[x, x * x]])[0][0]);

writePlotlyHtml(
  new URL('./02a_quadratic_plot.html', import.meta.url).pathname,
  'Quadratic Regression - house age vs price',
  [
    { type: 'scatter', mode: 'markers', name: 'training',
      x: data.age, y: data.price, marker: { color: 'magenta', symbol: 'star', size: 12 } },
    { type: 'scatter', mode: 'lines', name: 'quadratic fit',
      x: ageGrid, y: curve, line: { color: 'blue' } },
    { type: 'scatter', mode: 'markers', name: 'predicted',
      x: test.age, y: predictions, marker: { color: 'red', size: 10 } },
  ],
  { xaxis: { title: 'house age (yr)' }, yaxis: { title: 'price ($k)' } },
);

console.log('\nWrote 02a_quadratic_plot.html');

File: 01_ML/01_supervised/01_regression/02_nonlinear_reg/python/02a_quadratic.py
"""ML: Non-Linear Regression - Polynomial / Quadratic.

House age vs price: brand-new homes priced high, mid-age homes priced low,
heritage homes priced high again — a U-shape that a parabola fits cleanly.

    price = a * age^2 + b * age + c

Run:
    python 02a_quadratic.py
"""

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

PLOT_PATH = "02a_quadratic_plot.png"


def main() -> None:
    data = {
        "age":   [1,   5,   10,  15,  20,  25,  30,  40,  55,  70,  85,  100],
        "price": [780, 690, 600, 540, 510, 500, 510, 560, 660, 770, 880, 990],
    }
    test = {"age": [3, 22, 45, 90]}

    df = pd.DataFrame(data)
    print("training data (price in $k):")
    print(df.to_string(index=False))

    poly = PolynomialFeatures(degree=2, include_bias=False)
    X = poly.fit_transform(df[["age"]].values)
    reg = LinearRegression().fit(X, df["price"])

    # PolynomialFeatures returns columns in order [x, x^2]
    b, a = reg.coef_
    c = reg.intercept_
    print(f"\nfit: price = {a:.4f} * age^2 + {b:.4f} * age + {c:.4f}")

    tst = pd.DataFrame(test)
    tst["price"] = reg.predict(poly.transform(tst[["age"]].values))
    print("\npredictions:")
    print(tst.to_string(index=False))

    age_grid = np.linspace(df.age.min(), df.age.max(), 200)
    curve = reg.predict(poly.transform(age_grid.reshape(-1, 1)))

    plt.figure()
    plt.xlabel("house age (yr)")
    plt.ylabel("price ($k)")
    plt.scatter(df.age, df.price, color="magenta", marker="*", s=80, label="training")
    plt.plot(age_grid, curve, color="blue", label="quadratic fit")
    plt.scatter(tst.age, tst.price, color="red", label="predicted")
    plt.legend()
    plt.savefig(PLOT_PATH, bbox_inches="tight")
    print(f"\nWrote {PLOT_PATH}")


if __name__ == "__main__":
    main()
